core/src/main/java/org/testcontainers/dockerclient/DockerClientProviderStrategy.java

```diff
@@ -114,108 +115,33 @@ public static DockerClientProviderStrategy getFirstValidStrategy(List<DockerClie
-                        log.info("Found Docker environment with {}", strategy.getDescription());
-                        log.debug(
-                            "Transport type: '{}', Docker host: '{}'",
-                            TestcontainersConfiguration.getInstance().getTransportType(),
-                            strategy.getTransportConfig().getDockerHost()
-                        );
-
-                        log.debug("Checking Docker OS type for {}", strategy.getDescription());
-                        String osType = info.getOsType();
-                        if (StringUtils.isBlank(osType)) {
-                            log.warn("Could not determine Docker OS type");
-                        } else if (!osType.equals("linux")) {
-                            log.warn("{} is currently not supported", osType);
-                            throw new InvalidConfigurationException(osType + " containers are currently not supported");
-                        }
-
-                        if (strategy.isPersistable()) {
-                            TestcontainersConfiguration.getInstance().updateUserConfig("docker.client.strategy", strategy.getClass().getName());
-                        }
-
-                        return Stream.of(strategy);
-                    } catch (Exception | ExceptionInInitializerError | NoClassDefFoundError e) {
-                        @Nullable String throwableMessage = e.getMessage();
-                        @SuppressWarnings("ThrowableResultOfMethodCallIgnored")
-                        Throwable rootCause = Throwables.getRootCause(e);
-                        @Nullable String rootCauseMessage = rootCause.getMessage();
-
-                        String failureDescription;
-                        if (throwableMessage != null && throwableMessage.equals(rootCauseMessage)) {
-                            failureDescription = String.format("%s: failed with exception %s (%s)",
-                                    strategy.getClass().getSimpleName(),
-                                    e.getClass().getSimpleName(),
-                                    throwableMessage);
-                        } else {
-                            failureDescription = String.format("%s: failed with exception %s (%s). Root cause %s (%s)",
-                                    strategy.getClass().getSimpleName(),
-                                    e.getClass().getSimpleName(),
-                                    throwableMessage,
-                                    rootCause.getClass().getSimpleName(),
-                                    rootCauseMessage
-                            );
-                        }
-                        configurationFailures.add(failureDescription);
-
-                        log.debug(failureDescription);
-                        return Stream.empty();
-                    }
-                })
+                .flatMap(strategy -> tryOutStrategy(configurationFailures, strategy))
```

Comment:
This could also be `filter`, with `tryOutStrategy` returning a `boolean`, right?

Reply:
It could. I kept it like this, because I just made structural changes to the code and extracted methods.

Are you thinking about the `boolean` to make it more testable or to make the intent cleaner?

Reply:
Making the intent clearer was my initial thinking, but it would make the function slightly easier to test in isolation as well.